src/libraries/TwabLib.sol

```diff
@@ -18,435 +17,674 @@ import { ObservationLib, MAX_CARDINALITY } from "./ObservationLib.sol";
+        accountDetails
+      );
+
+      if (isNew) {
+        // If the index is new, then we increase the next index to use
+        accountDetails.nextObservationIndex = uint16(
+          RingBufferLib.nextIndex(uint256(index), MAX_CARDINALITY)
+        );
+
+        // Prevent the Account specific cardinality from exceeding the MAX_CARDINALITY.
+        // The ring buffer length is limited by MAX_CARDINALITY. IF the account.cardinality
+        // exceeds the max cardinality, new observations would be incorrectly set or the
+        // observation would be out of "bounds" of the ring buffer. Once reached the
+        // Account.cardinality will continue to be equal to max cardinality.
+        if (accountDetails.cardinality < MAX_CARDINALITY) {
+          accountDetails.cardinality += 1;
+        }
+      }
+
+      observation = ObservationLib.Observation({
+        balance: uint96(accountDetails.delegateBalance),
+        cumulativeBalance: _extrapolateFromBalance(newestObservation, currentTime),
+        timestamp: currentTime
+      });
+
+      // Write to storage
+      _account.observations[index] = observation;
+    }
+    // Write to storage
+    _account.details = accountDetails;
   }
 
   /**
-   * @notice Calculates the average balance held by a user for a given time frame.
-   * @dev Finds the average balance between start and end timestamp epochs.
-   *      Validates the supplied end time is within the range of elapsed time i.e. less then timestamp of now.
-   * @param _twabs          Individual user Observation recorded checkpoints passed as storage pointer
-   * @param _accountDetails User AccountDetails struct loaded in memory
-   * @param _startTime      Start of timestamp range as an epoch
-   * @param _endTime        End of timestamp range as an epoch
-   * @return uint256 Average balance of user held between epoch timestamps start and end
+   * @notice Looks up the oldest observation in the circular buffer.
+   * @param _observations The circular buffer of observations
+   * @param _accountDetails The account details to query with
+   * @return index The index of the oldest observation
+   * @return observation The oldest observation in the circular buffer
    */
-  function getAverageBalanceBetween(
-    ObservationLib.Observation[MAX_CARDINALITY] storage _twabs,
-    AccountDetails memory _accountDetails,
-    uint32 _startTime,
-    uint32 _endTime
-  ) internal view returns (uint256) {
-    uint32 _currentTime = uint32(block.timestamp);
-    _endTime = _endTime > _currentTime ? _currentTime : _endTime;
+  function getOldestObservation(
+    ObservationLib.Observation[MAX_CARDINALITY] storage _observations,
+    AccountDetails memory _accountDetails
+  ) internal view returns (uint16 index, ObservationLib.Observation memory observation) {
+    // If the TWAB is not initialized we go to the beginning of the TWAB circular buffer at index 0
+    if (_accountDetails.cardinality < MAX_CARDINALITY) {
```

Comment:
It's not because `cardinality` is lower than `MAX_CARDINALITY` that it means that the TWAB is not initialized no?
It's not initialized if `cardinality` is equal to `0` no?

Reply:
Yeah that's bad wording. It's `If the circular buffer has not been fully populated, we go to the beginning of the buffer at index 0.`

Reply:
Ah yes ok. Cause at index 0, it will be the oldest observation.